src/scheduler/MpiWorld.cpp

```diff
@@ -113,13 +185,37 @@ void MpiWorld::create(const faabric::Message& call, int newId, int newSize)
 
 void MpiWorld::destroy()
 {
-    // Destroy once per host
-    if (!isDestroyed.test_and_set()) {
-        shutdownThreadPool();
+    // Destroy once per thread the rank-specific data structures
+    // Remote message endpoints
+    if (mpiMessageEndpoints.size() > 0) {
+        for (auto& e : mpiMessageEndpoints) {
+            if (e != nullptr) {
+                e->close();
+            }
+        }
+        mpiMessageEndpoints.clear();
+    }
+
+    // Unacked message buffers
+    if (unackedMessageBuffers.size() > 0) {
+        for (auto& umb : unackedMessageBuffers) {
+            if (umb != nullptr) {
+                assert(umb->ids.empty() && umb->msgs.empty() &&
+                       umb->args.empty());
```

Comment:
There are a lot of assertions in this PR and it's not clear what role they're all performing. Let's discuss offline.